packages/vm/test/api/EIPs/eip-7002.spec.ts

```diff
@@ -0,0 +1,182 @@
+    20
+  )
+  const withdrawalsAddress = Address.fromString(bytesToHex(addressBytes))
+
+  return LegacyTransaction.fromTxData({
+    nonce,
+    gasPrice: BigInt(100),
+    data: concatBytes(validatorPubkey, amountBytes),
+    value: BigInt(1),
+    to: withdrawalsAddress,
+    gasLimit: 200_000,
+  }).sign(pkey)
+}
+
+describe('EIP-7002 tests', () => {
+  it('should correctly create requests', async () => {
+    const vm = await setupVM({ common })
+    const block = Block.fromBlockData(
+      {
+        header: {
+          number: 1,
+        },
+        transactions: [deploymentTx],
+      },
+      { common }
+    )
+    await vm.stateManager.putAccount(sender, acc)
+    await vm.stateManager.putAccount(addr, acc)
+
+    // Deploy withdrawals contract
+    const results = await vm.runBlock({
+      block,
+      skipHeaderValidation: true,
+      skipBlockValidation: true,
+      generate: true,
+    })
+
+    const root = results.stateRoot
+
+    const tx = generateTx(BigInt(0))
+
+    // Call withdrawals contract with a withdrawals request
+    const block2 = Block.fromBlockData(
+      {
+        header: {
+          number: 2,
+          parentBeaconBlockRoot: zeros(32),
+        },
+        transactions: [tx],
+      },
+      { common }
+    )
+
+    let generatedBlock: Block
+    vm.events.on('afterBlock', (e) => {
+      generatedBlock = e.block
+    })
+
+    await vm.runBlock({
+      block: block2,
+      skipHeaderValidation: true,
+      skipBlockValidation: true,
+      generate: true,
+    })
+
+    // Ensure the request is generated
+    assert.ok(generatedBlock!.requests!.length === 1)
+
+    const requestDecoded = RLP.decode(generatedBlock!.requests![0].bytes)
+
+    const sourceAddressRequest = requestDecoded[0] as Uint8Array
+    const validatorPubkeyRequest = requestDecoded[1] as Uint8Array
+    const amountRequest = requestDecoded[2] as Uint8Array
+
+    // Ensure the requests are correct
+    assert.ok(equalsBytes(sourceAddressRequest, tx.getSenderAddress().bytes))
+    assert.ok(equalsBytes(validatorPubkey, validatorPubkeyRequest))
+    assert.ok(equalsBytes(amountBytes, amountRequest))
+
+    await vm.runBlock({ block: generatedBlock!, skipHeaderValidation: true, root })
+
+    // Run block with 2 requests
+
+    const tx2 = generateTx(BigInt(1))
+    const tx3 = generateTx(BigInt(2))
+
+    const block3 = Block.fromBlockData(
+      {
+        header: {
+          number: 3,
+          parentBeaconBlockRoot: zeros(32),
+        },
+        transactions: [tx2, tx3],
+      },
+      { common }
+    )
+
+    await vm.runBlock({
+      block: block3,
+      skipHeaderValidation: true,
+      skipBlockValidation: true,
+      generate: true,
+    })
+
+    // Note: generatedBlock is now overridden with the new generated block (this is thus block number 3)
+    // Ensure there are 2 requests
+    assert.ok(generatedBlock!.requests!.length === 2)
```

Comment:
This is a fantastic test.  I didn't even think about being able catch the generated block from teh `afterBlock` event.

Reply:
Yes, actually I want to raise a point about this, isn't the entire point of `generate` to actually retrieve the generated block (i.e. build the block?). I was somewhat surprised that this block was not returned from `runBlock` so had to figure out if I would be able to extract it (if that was even possible). So this is something we could take a look at in the future (not part of this PR)